File: CodeBattleJava/src/main/java/clientlib/SampleSolver.java
package clientlib;

import java.util.*;

import static clientlib.Elements.*;


public class SampleSolver extends Solver {


    public List<Point> getPlayerTankCoordinates(){
            List<Point> playerTank = getCoordinates(TANK_DOWN, TANK_UP, TANK_LEFT, TANK_RIGHT);
            if(playerTank.size() == 0){
                playerTank.add(new Point(0,0));
            }
            return playerTank;
        }


    public List<Point> getOtherPlayersTanks(){
        List<Point> otherPlayers = getCoordinates(OTHER_TANK_DOWN, OTHER_TANK_UP, OTHER_TANK_LEFT, OTHER_TANK_RIGHT);
        return otherPlayers;
    }

    public List<Point> getBotsTanks(){
        List<Point> bots = getCoordinates(AI_TANK_DOWN, AI_TANK_UP, AI_TANK_LEFT, AI_TANK_RIGHT);
        return bots;
    }

    public List<Point> getBullets(){
        List<Point> bullets = getCoordinates(BULLET_UP, BULLET_DOWN, BULLET_LEFT, BULLET_RIGHT);
        return bullets;
    }

    public List<Point> getWormholes(){
        List<Point> wormholes = getCoordinates(WORM_HOLE);
        return wormholes;
    }

    public List<Point> getConstructions(){
        List<Point> constructions = getCoordinates(CONSTRUCTION);
        return constructions;
    }

    public List<Point> getDestroyedConstructions(){
        List<Point> constructions = getCoordinates(CONSTRUCTION_DESTROYED_DOWN,
                CONSTRUCTION_DESTROYED_UP,
                CONSTRUCTION_DESTROYED_LEFT,
                CONSTRUCTION_DESTROYED_RIGHT,
                CONSTRUCTION_DESTROYED,

                CONSTRUCTION_DESTROYED_DOWN_TWICE,
                CONSTRUCTION_DESTROYED_UP_TWICE,
                CONSTRUCTION_DESTROYED_LEFT_TWICE,
                CONSTRUCTION_DESTROYED_RIGHT_TWICE,

                CONSTRUCTION_DESTROYED_LEFT_RIGHT,
                CONSTRUCTION_DESTROYED_UP_DOWN,

                CONSTRUCTION_DESTROYED_UP_LEFT,
                CONSTRUCTION_DESTROYED_RIGHT_UP,
                CONSTRUCTION_DESTROYED_DOWN_LEFT,
                CONSTRUCTION_DESTROYED_DOWN_RIGHT);
        return constructions;
    }

    public List<Point> getBogs(){
        List<Point> bogs = getCoordinates(BOG);
        return bogs;
    }

    public List<Point> getSands() {
        List<Point> sand = getCoordinates(SAND);
        return sand;
    }

    public List<Point> getMoats() {
        List<Point> moat = getCoordinates(MOAT_HORIZONTAL, MOAT_VERTICAL);
        return moat;
    }

    public List<Point> getHedgehogs() {
        List<Point> hedgehog = getCoordinates(HEDGEHOG);
        return hedgehog;
    }

    public List<Point> getWalls(){
        List<Point> walls = getCoordinates(BATTLE_WALL);
        return walls;
    }

    public List<Point> getBarriers(){
        List<Point> barriers = new ArrayList<>();
        barriers.addAll(getWalls());
        barriers.addAll(getConstructions());
        barriers.addAll(getDestroyedConstructions());
        barriers.addAll(getOtherPlayersTanks());
        barriers.addAll(getBotsTanks());
        barriers.addAll(getHedgehogs());
        return barriers;
    }

    public List<Point> getAmmoBonuses(){
        List<Point> ammo = getCoordinates(BONUS_AMMO);
        return ammo;
    }

    public List<Point> getMedKitBonuses(){
        List<Point> medkit = getCoordinates(MEDKIT);
        return medkit;
    }

    public boolean isNear(int x, int y, Elements el){
        return isAt(x+1,y,el) ||
                isAt(x-1,y, el) ||
                isAt(x, y-1, el) ||
                isAt(x, y+1, el);
    }

    public boolean isBarrierAt(int x, int y){
        return getBarriers().contains(new Point(x,y));
    }

    public boolean isAnyOfAt(int x, int y, Elements... elements){
        boolean result = false;
        for (Elements el : elements){
            result = isAt(x, y, el);
            if(result) break;
        }
        return result;
    }

    public boolean isAt(int x, int y, Elements element){
        if(isOutOfBounds(x, y)){
            return false;
        } else{
            return field[x][y] == element;
        }
    }

    public int countNear(int x, int y, Elements element){
        int counter = 0;
        if(isAt(x+1, y, element)) counter++;
        if(isAt(x-1, y, element)) counter++;
        if(isAt(x, y+1, element)) counter++;
        if(isAt(x, y-1, element)) counter++;
        return counter;
    }

    public boolean isOutOfBounds(int x, int y){
        return x >= field.length || y >= field.length || x < 0 || y < 0;
    }


    public List<Point> getCoordinates(Elements... searchElements){
        Set<Elements> searchSetElements = new HashSet<>(Arrays.asList(searchElements));
        List<Point> elementsCoordinates = new ArrayList<>();

        for (int y = 0; y < field.length; y++) {
            for (int x = 0; x < field.length; x++) {
                if (searchSetElements.contains(field[x][y])) {
                    elementsCoordinates.add(new Point(x,y));
                }
            }
        }
        return elementsCoordinates;
    }






    @Override
    public String move() {
        return left();
    }

}

File: CodeBattleKotlin/src/main/kotlin/clientlib/SampleSolver.kt
package clientlib

import clientlib.Elements.*
import java.util.Scanner
import kotlin.math.sqrt

private val SIZE_OF_BOARD = sqrt(board.length.toDouble()).toInt()

class SampleSolver : Solver() {

    /**
     * This should be your bot's code.
     */
    override fun move(): String {
        val action: String
        val scanner = Scanner(System.`in`)
        action = scanner.nextLine()
        return when (action) {
            "w" -> up()
            "s" -> down()
            "a" -> left()
            "d" -> right()
            "q" -> act()
            else -> ""
        }
    }

    fun getPlayerTank(): List<Point> {
        return getCoordinatesPlayerTank(arrayOf(TANK_UP, TANK_RIGHT, TANK_DOWN, TANK_LEFT))
    }

    fun getOtherPlayersTanks(): List<Point> {
        return getCoordinatesElements(arrayOf(OTHER_TANK_UP, OTHER_TANK_RIGHT, OTHER_TANK_DOWN, OTHER_TANK_LEFT))

    }

    fun getBotsTanks(): List<Point> {
        return getCoordinatesElements(arrayOf(AI_TANK_UP, AI_TANK_RIGHT, AI_TANK_DOWN, AI_TANK_LEFT))
    }

    fun getBullets(): List<Point> {
        return getCoordinatesElements(arrayOf(BULLET, BULLET_UP, BULLET_RIGHT, BULLET_DOWN, BULLET_LEFT))
    }

    fun getWormholes(): List<Point> {
        return getCoordinatesPoint(board, WORM_HOLE)
    }

    fun getConstructions(): List<Point> {
        return getCoordinatesElements(arrayOf(CONSTRUCTION, CONSTRUCTION_DESTROYED_DOWN, CONSTRUCTION_DESTROYED_UP,
                CONSTRUCTION_DESTROYED_LEFT, CONSTRUCTION_DESTROYED_RIGHT, CONSTRUCTION_DESTROYED_DOWN_TWICE,
                CONSTRUCTION_DESTROYED_UP_TWICE, CONSTRUCTION_DESTROYED_LEFT_TWICE, CONSTRUCTION_DESTROYED_RIGHT_TWICE,
                CONSTRUCTION_DESTROYED_LEFT_RIGHT, CONSTRUCTION_DESTROYED_UP_DOWN, CONSTRUCTION_DESTROYED_UP_LEFT,
                CONSTRUCTION_DESTROYED_RIGHT_UP, CONSTRUCTION_DESTROYED_DOWN_LEFT, CONSTRUCTION_DESTROYED_DOWN_RIGHT))
    }

    fun getDestroyedConstructions(): List<Point> {
        return getCoordinatesPoint(board, CONSTRUCTION_DESTROYED)
    }

    fun getBogs(): List<Point> {
        return getCoordinatesPoint(board, BOG)
    }

    fun getSands(): List<Point> {
        return getCoordinatesPoint(board, SAND)
    }

    fun getMoats(): List<Point> {
        return getCoordinatesElements(arrayOf(MOAT_HORIZONTAL, MOAT_VERTICAL))
    }

    fun getHedgehogs(): List<Point> {
        return getCoordinatesPoint(board, HEDGEHOG)
    }

    fun getWalls(): List<Point> {
        return getCoordinatesPoint(board, BATTLE_WALL)
    }

    fun getBarriers(): List<Point> {
        return getCoordinatesElements(arrayOf(BATTLE_WALL, CONSTRUCTION, CONSTRUCTION_DESTROYED_DOWN,
                CONSTRUCTION_DESTROYED_UP, CONSTRUCTION_DESTROYED_LEFT, CONSTRUCTION_DESTROYED_RIGHT,
                CONSTRUCTION_DESTROYED_DOWN_TWICE, CONSTRUCTION_DESTROYED_UP_TWICE, CONSTRUCTION_DESTROYED_LEFT_TWICE,
                CONSTRUCTION_DESTROYED_RIGHT_TWICE, CONSTRUCTION_DESTROYED_LEFT_RIGHT, CONSTRUCTION_DESTROYED_UP_DOWN,
                CONSTRUCTION_DESTROYED_UP_LEFT, CONSTRUCTION_DESTROYED_RIGHT_UP, CONSTRUCTION_DESTROYED_DOWN_LEFT,
                CONSTRUCTION_DESTROYED_DOWN_RIGHT, OTHER_TANK_UP, OTHER_TANK_RIGHT, OTHER_TANK_DOWN, OTHER_TANK_LEFT,
                AI_TANK_UP, AI_TANK_RIGHT, AI_TANK_DOWN, AI_TANK_LEFT, HEDGEHOG))
    }

    fun getAmmoBonuses(): List<Point> {
        return getCoordinatesPoint(board, BONUS_AMMO)
    }

    fun getMedKitBonuses(): List<Point> {
        return getCoordinatesPoint(board, MEDKIT)
    }

    fun isAnyOfAt(x: Int, y: Int, elements: Array<Elements>): Boolean {
        for (element in elements) {
            if (getCoordinatesPoint(board, element).contains(Point(x, y))) {
                return true
            }
        }
        return false
    }

    fun isAt(x: Int, y: Int, element: Elements): Boolean {
        return getCoordinatesPoint(board, element).contains(Point(x, y))
    }

    fun isNear(x: Int, y: Int, element: Elements): Boolean {
        for (point in returnsArrayPointsAround(x, y)) {
            return (getCoordinatesPoint(board, element).contains(point))
        }
        return false
    }

    fun isBarrierAt(x: Int, y: Int): Boolean {
        return getBarriers().contains(Point(x, y))
    }

    fun countNear(x: Int, y: Int, element: Elements): Int {
        var count = 0
        for (point in returnsArrayPointsAround(x, y)) {
            if (getCoordinatesPoint(board, element).contains(point)) {
                count++
            }
        }
        return count
    }

    private fun getCoordinatesPoint(board: String, element: Elements): List<Point> {
        val listElements: MutableList<Point> = mutableListOf()
        var coordinateX: Int
        var coordinateY: Int
        for ((index, char) in board.withIndex()) {
            if (char == element.element) {
                coordinateX = (index + 1) % SIZE_OF_BOARD
                if (coordinateX == 0) {
                    coordinateX = SIZE_OF_BOARD
                }
                coordinateY = index / SIZE_OF_BOARD + 1
                listElements.add(Point(coordinateX, coordinateY))
            }
        }
        return listElements
    }

    private fun getCoordinatesPlayerTank(elements: Array<Elements>): List<Point> {
        for (element in elements) {
            if (board.contains(element.element)) {
                return getCoordinatesPoint(board, element)
            }
        }
        return listOf()
    }

    private fun getCoordinatesElements(elements: Array<Elements>): List<Point> {
        val listElements: MutableList<Point> = mutableListOf()
        for (element in elements) {
            listElements.addAll(getCoordinatesPoint(board, element))
        }
        return listElements
    }

    private fun returnsArrayPointsAround(x: Int, y: Int): Array<Point> {
        return arrayOf(Point(x, y - 1), Point(x + 1, y - 1), Point(x + 1, y), Point(x + 1, y + 1),
                Point(x, y + 1), Point(x - 1, y + 1), Point(x - 1, y), Point(x - 1, y - 1))
    }
}
